### ISSUE TITLE
Take a picture not working

### ISSUE THREAD

**Issue by duncan-dao-90** [alexk111/ngImgCrop#40]
On iPad or iPhone, when I choose "Take a photo" instead of "Use existing" to choose an image to crop,
the crop area will be blank and the image is not loaded.

Could you have a look at this?
Thanks.

**Comment by brianfeister** [comment]
+1, this is a huge problem for me with my mobile app ;)  

I've found two other directives (aside from compatibility I prefer `ngImgCrop`) that work properly on iOS:

http://jsbin.com/fovovu/1/edit?js,output (http://andyshora.com/angular-image-cropper.html)

http://think2011.github.io/localResizeIMG/ (https://github.com/think2011/localResizeIMG/blob/gh-pages/LocalResizeIMG.js)

**Comment by spectravp** [comment]
+1, huge for me also

**Comment by brianfeister** [comment]
From my above referenced alternate libraries, I wanted to call attention to 

```javascript
      if( navigator.userAgent.match(/iphone/i) ) {
        var mpImg = new MegaPixImage(item);
     }
```

It seems that there is a special method `MegaPixImage` for reading the data that mobile safari pushed to the DOM's `<input>`.  Perhaps we could add that?

**Comment by brianfeister** [comment]
Thoughts @alexk111?

**Comment by anhdd-savvycom** [comment]
Well I've firgured this out. Method "setNewImageSource" which checks for image rotation caused this. Add an onload handler fixed this problem.

**Comment by brianfeister** [comment]
@anhdd-savvycom can you post code?

**Comment by anhdd-savvycom** [comment]
@brianfeister I replaced method from line 1549 of compile/unminified/ng-img-crop.js with below:

    this.setNewImageSource=function(imageSource) {
        image=null;
        resetCropHost();
        events.trigger('image-updated');
        if(!!imageSource) {
            var newImage = new Image();
            if(imageSource.substring(0,4).toLowerCase()==='http') {
                newImage.crossOrigin = 'anonymous';
            }
            newImage.onload = function(){
                events.trigger('load-done');

                cropEXIF.getData(newImage,function(){
                    var orientation=cropEXIF.getTag(newImage,'Orientation');

                    if([3,6,8].indexOf(orientation)>-1) {
                        var canvas = document.createElement("canvas"),
                        ctx=canvas.getContext("2d"),
                        cw = newImage.width, ch = newImage.height, cx = 0, cy = 0, deg=0;
                        switch(orientation) {
                            case 3:
                                cx=-newImage.width;
                                cy=-newImage.height;
                                deg=180;
                                break;
                            case 6:
                                cw = newImage.height;
                                ch = newImage.width;
                                cy=-newImage.height;
                                deg=90;
                                break;
                            case 8:
                                cw = newImage.height;
                                ch = newImage.width;
                                cx=-newImage.width;
                                deg=270;
                            break;
                        }

                        canvas.width = cw;
                        canvas.height = ch;
                        ctx.rotate(deg*Math.PI/180);
                        ctx.drawImage(newImage, cx, cy);

                        image=new Image();
                        image.onload = function() {
                            resetCropHost();
                            events.trigger('image-updated');
                        };

                        image.src = canvas.toDataURL("image/png");
                    } else {
                        image=newImage;
                        resetCropHost();
                        events.trigger('image-updated');
                    }
                });
            };
            newImage.onerror=function() {
                events.trigger('load-error');
            };
            events.trigger('load-start');
            newImage.src=imageSource;
        }
    };

Hope this help.

**Comment by mike-suggitt** [comment]
You can't rely on onload alone to trigger as it never fires for cached images. see something like this -> http://mikefowler.me/2014/04/22/cached-images-load-event/

**Comment by brianfeister** [comment]
@mike-suggitt to put it in context though, we're talking about images captured via iOS "take a picture / video" for an input field.  Since those images are generated at the time of capture, there is no way they would ever be cached.  I'll be the first one to admit that I don't understand all of the moving parts here. My goal is just to get iOS file input fields to accept camera captures (since I'm working on a mobile app ;)

**Comment by mike-suggitt** [comment]
@brianfeister fair enough but probably best not to change a global method. If the cause of the problem is the image rotation work then maybe pick that out? I currently use ngImgCrop without that feature set and it works fine on a mobile (bar the image rotation obviously!)

**Comment by spectravp** [comment]
@mike-suggitt I agree. It's not a solution if it creates a new problem in its place.

**Comment by brianfeister** [comment]
@mike-suggitt I'm not seeing an option to disable rotation, where is that?

**Comment by danieljnewsome** [comment]
I'm still having issues even with the fix from @anhdd-savvycom above... tried a couple of things... Does anyone have a js file that's working for them?

**Comment by mike-suggitt** [comment]
sorry @brianfeister I wasn't clear. I simply didn't include the commit related to image rotation in my code base. You should be able to pick out the code easily enough though and comment it out/delete it. Just look at the diff to see where the offending code sits.

**Comment by anthonyhoman** [comment]
anhdd-savvycom's fix worked for me.  But that means I'm locally modifying the source file in the meantime.  Any word from alexk111 on whether or not he will include this fix officially?

**Comment by danieljnewsome** [comment]
For me, 0.3.2, on an iPhone 6 would pull up a picture from the camera roll properly, but failed on taking a picture (blank picture, unresponsive afterwards). 0.3.1 would take a picture or pull one from camera roll, but did not succeed in calculating EXIF.  I tried putting the EXIF code in 0.3.1, but it did not work... which makes me think that EXIF code works fine on desktop, but not in iPhone app.   BTW, it was a hybrid app, built with Phonegap, so not Chrome

**Comment by anthonyhoman** [comment]
Looks like I spoke too soon.  anhdd-savvycom's fix isn't working for me.  "Take a Picture" is working on Chrome on the iPad for me, but Safari does not work.

**Comment by dnlmzw** [comment]
+1 - I'm experiencing this on an iPhone 5S 8.1 in both Safari and Chrome. Also I'm experiencing that some images from the photo gallery load, while others don't. My theory is that it has something to do with filesize - but I can't say for sure since I haven't tested it yet.

**Comment by hbbh** [comment]
I believe the issue is to do with the way the images are manipulated when EXIF data exists. There is a call to .toDataURL("image/png") from the HTML5 canvas that doesn't work if the image taken by the camera is too big. I have just fixed the issue within iblank's fork of this project, which also has a bunch of lovely stuff in it, including fixed aspect ratios.

The pull request is here. https://github.com/iblank/ngImgCrop/pull/1

I have created a GIST (untested) for those of you using the unforked version of the lib.

https://gist.github.com/hbbh/d5f21c4120665f9429e3

this is a replacement for the setNewImageSource function (as per @anhdd-savvycom comment above), which is originally found in crop-host.js file and should just replace the same function in the compilled version. I haven't created a pull request as I've not had time to test in the original repo.

**Comment by ilianaza** [comment]
+1 Awesome library, but not so useful without the mobile support. It would be by far the best image cropping library if it is mobile compliant.

**Comment by kevinsuh** [comment]
any official word on this from alexk111?

**Comment by bzmillerboy** [comment]
This issue could be solved by adding a rotate button to the UI. If you guys agree this is the proper approach, we'll create a PR to add the enhancement.

**Comment by rgins16** [comment]
Is there a definitive solution for this issue? I am experiencing it right now on my iPhone 5C.

**Comment by nitheeshskiploop** [comment]
Any update on this issue ? I am facing the same and changes in file ng-img-crop.js didn't work for me.

**Comment by EternallLight** [comment]
Changes from @anhdd-savvycom did work for me.

**Comment by abhijit-dat** [comment]
I have captured an image by using webcam. But did not get the right output, following was my code
//  $scope.capturedImageUrl has dataUrl of the image generaed from canvas

        $scope.myImage='';
        $scope.myCroppedImage='';
        var file = $scope.capturedImageUrl;
        var reader = new FileReader();
        $scope.$apply(function($scope){
            $scope.myImage = $scope.capturedImageUrl;
        });
        reader.readAsDataURL(file);

Returns error:  "$apply already in progress"
If I remove $scope.$apply 
Returns error: TypeError: Failed to execute 'readAsDataURL' on 'FileReader': parameter 1 is not of type 'Blob'.
Am I missing something? Or doing totally wrong.

**Comment by Partho-paul** [comment]
`   <img-crop image="myImage" result-image="myCroppedImage" result-image-quality="1.0" change-on-fly="true" result-image-size="50"  @*on-load-begin="startspinner()" on-load-done="closetimer()"*@></img-crop>`


Just remove the CallBacks : Onloadstart and onloadstop and things might be fine...
